### PR TITLE
xds: allow injecting bootstrapOverride in xdsNameResolverProvider

### PR DESCRIPTION
**Pull request by YifeiZhuang**
fix https://github.com/grpc/grpc-java/issues/7819

### CODE REVIEW

File: xds/src/main/java/io/grpc/xds/XdsNameResolverProvider.java [hunk 1]
```diff
@@ -43,8 +43,11 @@
 
   private static final String SCHEME = "xds";
 
-  @Override
-  public XdsNameResolver newNameResolver(URI targetUri, Args args) {
+  /**
+   * Allows injecting bootstrapOverride to the name resolver.
+   * */
+  public XdsNameResolver newNameResolver(URI targetUri, Args args,
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsNameResolverProvider.java:49]: Who would call this method? I think we want `bootstrapOverride` to be passed to a constructor/factoryMethod for the provider itself.

Comment (YifeiZhuang) [reply xds/src/main/java/io/grpc/xds/XdsNameResolverProvider.java:49]: Right, i didn't test it. 
The `bootstrapOverride` needs to be saved here and then passed to `xdsNameResolver` when channel creates a new resolver.

File: xds/src/main/java/io/grpc/xds/XdsNameResolver.java [hunk 2]
```diff
@@ -124,22 +124,24 @@
...
       SynchronizationContext syncContext, ScheduledExecutorService scheduler,
       XdsClientPoolFactory xdsClientPoolFactory, ThreadSafeRandom random,
-      FilterRegistry filterRegistry) {
+      FilterRegistry filterRegistry, @Nullable Map<String, ?> bootstrapOverride) {
     authority = GrpcUtil.checkAuthority(checkNotNull(name, "name"));
     this.serviceConfigParser = checkNotNull(serviceConfigParser, "serviceConfigParser");
     this.syncContext = checkNotNull(syncContext, "syncContext");
     this.scheduler = checkNotNull(scheduler, "scheduler");
     this.xdsClientPoolFactory = checkNotNull(xdsClientPoolFactory, "xdsClientPoolFactory");
+    this.xdsClientPoolFactory.setBootstrapOverride(bootstrapOverride);
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsNameResolver.java:144]: This mutates the default factory, which would impact other xds clients. Create a new instance if bootstrapOverride is set.

File: xds/src/main/java/io/grpc/xds/XdsNameResolverProvider.java [hunk 3]
```diff
@@ -41,11 +41,24 @@
...
+  private Map<String, ?> bootstrapOverride;
+
+  /**
+   * A convenient method to create a {@link XdsNameResolverProvider} with custom scheme and
+   * bootstrap.
+   */
+  public static XdsNameResolverProvider createForTest(String scheme,
+                                               @Nullable Map<String, ?> bootstrapOverride) {
+    XdsNameResolverProvider provider = new XdsNameResolverProvider();
+    provider.scheme = checkNotNull(scheme, "scheme");
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsNameResolverProvider.java:54]: Add a constructor? That'd allow the fields to be final. You'd need to make a no-arg constructor as well.

File: xds/src/main/java/io/grpc/xds/XdsNameResolverProvider.java [hunk 4]
```diff
@@ -41,19 +41,22 @@
...
+    this(SCHEME, null);
+  }
 
   /**
-   * A convenient method to create a {@link XdsNameResolverProvider} with custom scheme and
-   * bootstrap.
+   * A convenient constructor to allow creating a {@link XdsNameResolverProvider} with custom scheme
+   * and bootstrap.
    */
-  public static XdsNameResolverProvider createForTest(String scheme,
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsNameResolverProvider.java]: I had intended for the constructor to be private, and for us to still have the "for test" factory method to discourage it for other uses.

**Review by ejona86 — APPROVED** [review]
Just a change to restore the static factory method.

**Pull request merged**